Revised the quickstart guide and added separate react-router guide (#518)
clerk/clerk-docs: docs/quickstarts/react.mdx

Reviewer: I see we have [React Quickstart Repo](https://github.com/clerk/clerk-react-quickstart/tree/main) and the branch [Integrate Clerk with React](https://github.com/clerk/clerk-react-quickstart/tree/integrate-clerk-with-react) and we don't explain the differences between the two/what they can be used for.

Author: I think we have the same with the Next QS repos.

Reviewer: ahh, so if our docs don't indicate something about how to use the branches, the repo README's defiiiiniiitely should. because how else do user's even know about them 🤔 I'll reach out to marketing team to see how we want to move forward

Author: I think the READMEs were mostly from @domitriusclark, so tagging him for feedback.

  ```
</CodeBlockTabs>

### Set environment keys

In your React project's root folder, you may have an `.env` file alongside `package.json` and other configuration files. If you don't see it, create it.

Add the following code to your `.env.local` file to set your public and secret keys.

**Pro tip!** If you are signed into your [Clerk Dashboard](https://dashboard.clerk.com/), your keys should become visible by clicking on the eye icon.
<InjectKeys>

```sh filename=".env"
VITE_CLERK_PUBLISHABLE_KEY={{pub_key}}
```

</InjectKeys>

### Wrap your app in `<ClerkProvider />`

The [`<ClerkProvider />`](/docs/components/clerk-provider) component provides active session and user context to Clerk's hooks and other components. Import it into your app by adding `import { ClerkProvider } from '@clerk/clerk-react'` at the top of your `/src/main.tsx` file. You will need to import your Publishable Key and provide it to your `<ClerkProvider />` using the `publishableKey` prop. 

```tsx filename="src/main.tsx" {5, 7, 9-11, 15, 17}
import React from 'react'
import ReactDOM from 'react-dom/client'
import App from './App.tsx'
import './index.css'
import { ClerkProvider } from '@clerk/clerk-react'

> const PUBLISHABLE_KEY = import.meta.env.VITE_CLERK_PUBLISHABLE_KEY

Reviewer: ```suggestion
// Import your publishable key
const PUBLISHABLE_KEY = import.meta.env.VITE_CLERK_PUBLISHABLE_KEY
```

description: Learn how to use Clerk to quickly and easily add secure authentication and user management to your React application.
---

<TutorialHero 
  quickstart="react"

>   exampleRepo={[ 

Reviewer: I can make example Repo accept more than one repositories, and update the design to follow.
This way, we can include the "React Quickstart Repo" as well as "React + TypeScript Quickstart Repo"

Author: "React + TypeScript Quickstart Repo" - if this just an example name you came up with? I don't think this is a thing.

Reviewer: oh, sorry, I saw a branch on that repo that looked like it was meant to serve as the repository but typed - this one: https://github.com/clerk/clerk-react-quickstart/tree/react-typescript-with-vite

Author: Gotcha. That branch is just the result of running the vite command to create the app. I don't think we need to link to it.

Basically the repo is:
- `main`, which has the `react-typescript-with-vite` and `integrate-clerk-with-react` merged into it.
- `integrate-react-router-dom-using-data-router-method` which is for the router guide.

To that end I think a link to the repo for the QS works and then a link to the branch for the Router guide works. Maybe something about merging the branch for the latter guide.

    {
      title: "React Quickstart Repo",
      link: "https://github.com/clerk/clerk-react-quickstart"
      
    }
  ]}
  beforeYouStart={[
    {
      title: "Set up a Clerk application",
      link: "https://clerk.com/docs/quickstarts/setup-clerk"
    },
    {
      title: "Create a React application using Vite",
      link: "https://vitejs.dev/guide/#scaffolding-your-first-vite-project"
    }
  ]}
>

> - Use Clerk with React
> - Install `@clerk/clerk-react`
> - Set environment keys
> - Wrap your app in `<ClerkProvider />`
> - Embed the `<UserButton />` and `<SignInButton />`

Reviewer: ```suggestion
- Set up a React app with Vite
- Install `@clerk/clerk-react`
- Set environment keys
- Wrap your React app in `<ClerkProvider />`
- Embed the `<UserButton />` and `<SignInButton />`
```

> - Embed the `<UserButton />` and `<SignInButton />`

Author: We use <UserButton /> in the Next QS. React doesn't have middleware, so we don't get the automatic redirect to the Account Portal and need to provide an option. <SignInButton /> gives us that.

</TutorialHero>

> # Use Clerk with React

Author: Done. Just missed removing them from the old doc.

Learn how to use Clerk to quickly and easily add secure authentication and user management to your React application.

<Steps>

> ### Install React using Vite

Reviewer: ```suggestion
### Set up a React application using Vite
```

Reviewer: If we have this section on how to create a react app with vite, then it probably shouldn't be in the "prerequisites", because it's actually a step in the tutorial - not a step before.

> We don't recommend using `create-react-app` for scaffoling your new React application. Instead we recommend using Vite to setup your application. 

Reviewer: ```suggestion
Scaffold your new React application using [Vite](https://vitejs.dev/guide/#scaffolding-your-first-vite-project).
```

> <CodeBlockTabs type="installer" options={["npm", "pnpm", "yarn"]}>

Reviewer: So everywhere else in the docs, we have it npm -> yarn -> pnpm
Is there a reason the order changed? Was it discussed? We should keep it the same across the docs for consistency, so whatever we decide here needs to be reflected everywhere else and vice versa.

Author: I just added it that way. I'm not strongly attached to this.

  ```bash filename="terminal"
  npm create vite@latest clerk-react -- --template react-ts
  cd clerk-react
  npm install
  npm run dev

  ```

  ```bash filename="terminal"
  pnpm create vite clerk-react --template react-ts
  cd clerk-react
  pnpm install
  pnpm dev
  ```

  ```bash filename="terminal"
  yarn create vite clerk-react--template react-ts
  cd clerk-react
  yarn install
  yarn dev
  ```
</CodeBlockTabs>

### Install `@clerk/clerk-react`

> Clerk's React SDK gives you access to prebuilt components, hooks, and helpers for React. Install it by running the following command in your terminal:

Reviewer: ```suggestion
Clerk's React SDK gives you access to prebuilt [components](/docs/components/overview), [hooks](/docs/references/react/use-user), and helpers for React. Install it by running the following command in your terminal:
```

<InjectKeys>

```sh filename=".env"
VITE_CLERK_PUBLISHABLE_KEY={{pub_key}}
```

</InjectKeys>

### Wrap your app in `<ClerkProvider />`

The [`<ClerkProvider />`](/docs/components/clerk-provider) component provides active session and user context to Clerk's hooks and other components. Import it into your app by adding `import { ClerkProvider } from '@clerk/clerk-react'` at the top of your file. You will need to import your Publishable Key and provide it to your `<ClerkProvider />` using the `publishableKey` prop. 

```tsx filename="src/main.tsx" {5, 7, 9-11, 15, 17}
import React from 'react'
import ReactDOM from 'react-dom/client'
import App from './App.tsx'
import './index.css'
import { ClerkProvider } from '@clerk/clerk-react'

const PUBLISHABLE_KEY = import.meta.env.VITE_CLERK_PUBLISHABLE_KEY

if (!PUBLISHABLE_KEY) {
  throw new Error("Missing Publishable Key")
}

ReactDOM.createRoot(document.getElementById('root')!).render(
  <React.StrictMode>
    <ClerkProvider publishableKey={PUBLISHABLE_KEY}>
      <App />
    </ClerkProvider>
  </React.StrictMode>,

> )

Reviewer: This information should be included somewhere. 
Also, I used the React starter and the port is 5173; I think that's a vite thing. But we need to make sure we have the React "starter" updated before this docs tutorial is released - it needs to match this tutorial exactly

```suggestion
)

At this point, your application should use Clerk's [Account Portal](/docs/account-portal/overview) pages that are available out of box. Start your React application via `npm run dev`, visit `http://localhost:5173`, and sign up to get access to your application. 
```

Author: Yeah, 5173 is a Vite thing. Good addition.

```
### Embed the `<UserButton />` and `<SignInButton />`

Clerk offers a set of [prebuilt components](/docs/components/overview) to add functionality to your app with minimal effort. The [`<UserButton />`](/docs/components/user/user-button) allows users to manage their account information and log out, completing the full authentication circle.The [`<SignInButton />`](/docs/components/unstyled/sign-in-button) is an unstyled component that you can use to either create a basic button or to wrap around your styled button and turn it into a Sign In button.

Lastly you will use the [`<SignedIn />`](/docs/components/control/signed-in) and [`<SignedOut />`](/docs/components/control/signed-out) control components. These will allow you to render the `<UserButton />` only to users who are signed in and `<SignInButton />` only to users who are signed out.

> Add the following anywhere inside `<ClerkProvider />` in your app. First add `import { UserButton, SignInButton, SignedIn, SignedOut } from "@clerk/clerk-react";` to the top of your file. Then add the following somewhere inside of your application -- a header area is a great place. The `afterSignOutUrl` prop lets you customize what page the user will be redirected to after sign out.

Reviewer: You mention a header area, but the code example doesn't have anything about a header area. The code is simply embedded in a div in the App()

Author: Its a reference to a common practice. I can remove it though.  I'll trim that down.

```tsx filename="src/App.tsx {1, 7-12}
import { SignInButton, SignedIn, SignedOut, UserButton } from "@clerk/clerk-react"

function App() {

  return (
    <div>
      <SignedOut>
        <SignInButton />
      </SignedOut>
      <SignedIn>
        <UserButton afterSignOutUrl="/" />
      </SignedIn>
    </div>
  )
}

export default App
```

> If you removed the Vite boilerplate from `App.tsx` by copying and pasting the boce, you probably want to also delete `src/App.css` and `src/assets/react.svg`.

Reviewer: Not sure if this is even necessary in the quickstart - the users can decide how they want to clean up/maintain their app. This isn't something that affects how Clerk gets up and running for this tutorial.

Author: I think its worth the mention. I'm fine with removing it though.

<InjectKeys>

```sh filename=".env"
VITE_CLERK_PUBLISHABLE_KEY={{pub_key}}
```

</InjectKeys>

### Wrap your app in `<ClerkProvider />`

> The [`<ClerkProvider />`](/docs/components/clerk-provider) component provides active session and user context to Clerk's hooks and other components. Import it into your app by adding `import { ClerkProvider } from '@clerk/clerk-react'` at the top of your file. You will need to import your Publishable Key and provide it to your `<ClerkProvider />` using the `publishableKey` prop. 

Author: The only issue would be whether users tend to save after each step. If so they will get an error after step. Aside from that question, I'm happy to make this change.

> The [`<ClerkProvider />`](/docs/components/clerk-provider) component provides active session and user context to Clerk's hooks and other components. Import it into your app by adding `import { ClerkProvider } from '@clerk/clerk-react'` at the top of your file. You will need to import your Publishable Key and provide it to your `<ClerkProvider />` using the `publishableKey` prop. 

Author: Maybe this, to break it down a little but also avoid the error:

1. import into your app...
2. define publishable key
3. import the publishable key  
4. Wrap <App> in <ClerkProvider>

> The [`<ClerkProvider />`](/docs/components/clerk-provider) component provides active session and user context to Clerk's hooks and other components. Import it into your app by adding `import { ClerkProvider } from '@clerk/clerk-react'` at the top of your file. You will need to import your Publishable Key and provide it to your `<ClerkProvider />` using the `publishableKey` prop. 

Author: I've updated this in line with my last suggestion -- breaking it up some but trying to avoid errors from saves/tests by the user between steps.

```tsx filename="src/main.tsx" {5, 7, 9-11, 15, 17}
import React from 'react'
import ReactDOM from 'react-dom/client'
import App from './App.tsx'
import './index.css'
import { ClerkProvider } from '@clerk/clerk-react'

const PUBLISHABLE_KEY = import.meta.env.VITE_CLERK_PUBLISHABLE_KEY

if (!PUBLISHABLE_KEY) {
  throw new Error("Missing Publishable Key")
}

ReactDOM.createRoot(document.getElementById('root')!).render(
  <React.StrictMode>
    <ClerkProvider publishableKey={PUBLISHABLE_KEY}>
      <App />
    </ClerkProvider>
  </React.StrictMode>,
)

At this point, your application should use Clerk's [Account Portal](/docs/account-portal/overview) pages that are available out of box. Start your React application via `npm run dev`, visit `http://localhost:5173`, and sign up to get access to your application. 

> ```

Author: These were in the wrong spot. Moved them to the correct location.

  return (
    <ClerkProvider publishableKey={clerkPubKey}>
      <div>Hello from clerk</div>
ReactDOM.createRoot(document.getElementById('root')!).render(
  <React.StrictMode>
    <ClerkProvider publishableKey={PUBLISHABLE_KEY}>
      <App />
    </ClerkProvider>
  );
}
  </React.StrictMode>,
)

export default App;
At this point, your application should use Clerk's [Account Portal](/docs/account-portal/overview) pages that are available out of box. Start your React application via `npm run dev`, visit `http://localhost:5173`, and sign up to get access to your application. 
```
### Embed the `<UserButton />` and `<SignInButton />`

> ### Protecting your pages

Author: Are we keeping the logos and the counter? We removed all Next logos/copy from that repo and quickstart so I followed that pattern here.

I added a couple of sentences for now. Once we're sure of this content I will update the repo.
